casper/src/main/resources/RevVault.rho

```diff
@@ -175,100 +219,139 @@ in {
+                              "amount" : amount,
+                              "block"  : { "number" : blockNumber, "timestamp" : timestamp, "sender" : sender }
+                            }
+                          ) {
+                            entryTuple => {
+                              new
+                                fromLogRet,
+                                toLogRet
+                              in {
+                                // Lookup vault associated with `toRevAddress`
+                                TreeHashMap!("get", vaultMap, toRevAddress, *toLogRet) |
+                                // Lookup log channel associated with `fromRevAddress`
+                                TreeHashMap!("get", vaultMap, fromRevAddress, *fromLogRet) |
+                                for (@(_, _, *fromLogCh) <- fromLogRet;
+                                     @(_, _, *toLogCh)   <- toLogRet) {
+                                  // Put the entry in both logs and return `result` on `ret`
+                                  fromLogCh!(entryTuple) |
+                                  toLogCh!(entryTuple) |
+                                  ret!(result)
+                                }
+                              }
+                            }
                           }
                         }
                       }
                     }
+                  } else {
+                    // Unsuccessful transfers are not logged
+                    ret!(result)
                   }
                 }
-              } |
-              // Set to Nil to disable logging
-              contract revVault(@"setLog", logCh, authKey, ret) = {
-                new authKeyValidCh in {
-                  @AuthKey!("check", *authKey, (*_revVault, ownRevAddress), *authKeyValidCh) |
-                  for (@result <- authKeyValidCh) {
-                    if (result) {
-                      new ack in {
-                        for (_ <- logStore) {
-                          logStore!(*logCh) |
-                          // Set logging on the purse so that we also see deposits into the vault
-                          purse!("setLog", *logCh, *ack) |
-                          for (_ <- ack) {
-                            ret!(true)
-                          }
-                        }
-                      }
-                    } else {
-                      ret!(false)
-                    }
+              }
+            } |
+            // Returns an entry of the vault's log
+            contract revVault(@"readLogEntry", ret) = {
+              new thmRet in {
+                // Lookup the log channel associated with `fromRevAddress`
+                TreeHashMap!("get", vaultMap, fromRevAddress, *thmRet) |
+                for (@(_, _, *logCh) <- thmRet) {
+                  // Read a transfer entry and return it on `ret`
+                  for (@("transfer", entry) <<- logCh) {
```

Comment:
I just want to say that `peek` would just consume **one** record on the `logCh` if `logCh` got some many of logs of it

Reply:
You are right. I'm not sure that we can _extract_ all log entries inside Rholang but only with direct _get data at name_ API method.